server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/DownloadRoutes.scala

```diff
@@ -0,0 +1,240 @@
+    SMono(authenticator.authenticate(request))
+      .flatMap((mailboxSession: MailboxSession) =>
+        SMono.fromTry(BlobId.of(request.param(blobIdParam)))
+          .flatMap(blobResolvers.resolve(_, mailboxSession))
+          .flatMap(blob => downloadBlob(
+            optionalName = queryParam(request, nameParam),
+            response = response,
+            blobContentType = queryParam(request, contentTypeParam)
+              .map(ContentType.of)
+              .getOrElse(blob.contentType),
+            blob = blob))
+          .`then`())
+      .onErrorResume {
+        case e: UnauthorizedException => SMono.fromPublisher(handleAuthenticationFailure(response, LOGGER, e)).`then`()
+        case _: BlobNotFoundException => SMono.fromPublisher(response.status(SC_NOT_FOUND).send).`then`()
+        case e =>
+          LOGGER.error("Unexpected error", e)
+          SMono.fromPublisher(handleInternalError(response, LOGGER, e)).`then`()
+      }
+      .subscribeOn(Schedulers.elastic)
+      .asJava()
+      .`then`()
+
+  private def downloadBlob(optionalName: Option[String],
+                           response: HttpServerResponse,
+                           blobContentType: ContentType,
+                           blob: Blob): SMono[Unit] =
+    SMono.fromPublisher(Mono.using(
+      () => blob.content,
+      (stream: InputStream) => addContentDispositionHeader(optionalName)
+        .compose(addContentLengthHeader(blob.size))
+        .apply(response)
+        .header(CONTENT_TYPE, blobContentType.asString)
+        .status(OK)
+        .send(ReactorUtils.toChunks(stream, BUFFER_SIZE)
+          .map(Unpooled.wrappedBuffer(_))
+          .subscribeOn(Schedulers.elastic))
+        .`then`(),
+      asJavaConsumer[InputStream]((stream: InputStream) => stream.close())))
+      .`then`()
+
+  private def addContentDispositionHeader(optionalName: Option[String]): HttpServerResponse => HttpServerResponse =
+    resp => optionalName.map(addContentDispositionHeaderRegardingEncoding(_, resp))
+      .getOrElse(resp)
+
+  private def addContentLengthHeader(sizeTry: Try[Size]): HttpServerResponse => HttpServerResponse =
+    resp => sizeTry
+      .map(size => resp.header("Content-Length", size.value.toString))
+      .getOrElse(resp)
+
+  private def addContentDispositionHeaderRegardingEncoding(name: String, resp: HttpServerResponse): HttpServerResponse =
+    if (CharMatcher.ascii.matchesAllOf(name)) {
+      resp.header("Content-Disposition", "attachment; filename=\"" + name + "\"")
+    } else {
+      resp.header("Content-Disposition", "attachment; filename*=\"" + EncoderUtil.encodeEncodedWord(name, Usage.TEXT_TOKEN) + "\"")
+    }
+
+  private def queryParam(httpRequest: HttpServerRequest, parameterName: String): Option[String] =
+    queryParam(parameterName, httpRequest.uri)
+
+  private def queryParam(parameterName: String, uri: String): Option[String] =
+    Option(new QueryStringDecoder(uri).parameters.get(parameterName))
```

Comment:
I think it could be simplified like this :  
`new QueryStringDecoder(uri).parameters.get(parameterName)
.asScala
.headOption`

Reply:
`new QueryStringDecoder(uri).parameters.get(parameterName)` can be null

Reply:
ok then